### ISSUE TITLE
Subsonic get api compatibility.

### ISSUE THREAD

**Issue by KimTheFirst** [ampache/ampache#1729]
# Environment

 * Ampache version: 3.8.8-develop
 * Web server + version: Apache2.4 or nginx 1.14, I've tested with both as primary server. Current configuration is apache behind nginx proxy/ssl termination.
 * Server operating system: Debian 9
 * Client operating system: Ubuntu 15 arm (ubuntu touch)

   - SubSonic. 
Specifcally I am trying to use this: https://github.com/inuitwallet/psub.
But the issue also appears to affect this: https://github.com/YunoHost-Apps/ampache_ynh/issues/14

# Settings

config_version = 34
web_path = ""
database_hostname = "localhost"
database_port = ""
database_name = "ampache"
database_username = "ampache"
database_password = "password"
secret_key = "key"
session_length = 3600
stream_length = 7200
remember_length = 604800
session_name = ampache
session_cookielife = 0
session_cookiesecure = 1
auth_methods = "mysql"
access_control	= "true"
require_session = "true"
require_localnet_session = "true"
getid3_tag_order = "id3v2,id3v1,vorbiscomment,quicktime,matroska,ape,asf,avi,mpeg,riff"
metadata_order = "getID3,filename"
metadata_order_video = "filename,getID3"
deferred_ext_metadata = "true"
additional_genre_delimiters = "[/]{2}|[/|\\\\|\|,|;]"
catalog_file_pattern = "mp3|mpc|m4p|m4a|aac|ogg|oga|wav|aif|aiff|rm|wma|asf|flac|opus|spx|ra|ape|shn|wv"
catalog_video_pattern = "avi|mpg|mpeg|flv|m4v|mp4|webm|mkv|wmv|ogv|mov|divx|m2ts"
catalog_playlist_pattern = "m3u|m3u8|pls|asx|xspf"
catalog_prefix_pattern = "The|An|A|Die|Das|Ein|Eine|Les|Le|La"
use_auth = "true"
default_auth_level = "guest"
ratings = "true"
userflags = "true"
directplay = "true"
sociable = "true"
licensing = "false"
art_order = "db,tags,folder,musicbrainz,lastfm,google"
lastfm_api_key = "d5df942424c71b754e54ce1832505ae2"
lastfm_api_secret = ""
wanted = "true"
wanted_types = "album,official"
channel = "true"
live_stream = "true"
refresh_limit = "60"
show_footer_statistics = "true"
debug = "true"
debug_level = 5
log_path = "/var/log/ampache"
log_filename = "%name.%Y%m%d.log"
site_charset = "UTF-8"
allow_public_registration = "false"
cookie_disclaimer = "false"
registration_display_fields = "fullname,website"
registration_mandatory_fields = "fullname"
transcode_player_customize = "true"
transcode_input = "-i %FILE%"
encode_args_mp3 = "-vn -b:a %BITRATE%K -c:a libmp3lame -f mp3 pipe:1"
encode_args_ogg = "-vn -b:a %BITRATE%K -c:a libvorbis -f ogg pipe:1"
encode_args_opus = "-vn -b:a %BITRATE%K -c:a libopus -compression_level 10 -vsync 2 -f ogg pipe:1"
encode_args_m4a = "-vn -b:a %BITRATE%K -c:a libfdk_aac -f adts pipe:1"
encode_args_wav = "-vn -b:a %BITRATE%K -c:a pcm_s16le -f wav pipe:1"
encode_args_flv = "-b:a %BITRATE%K -ar 44100 -ac 2 -v 0 -f flv -c:v libx264 -preset superfast -threads 0 pipe:1"
encode_args_webm = "-q %QUALITY% -f webm -c:v libvpx -maxrate %MAXBITRATE%k -preset superfast -threads 0 pipe:1"
encode_args_ts = "-q %QUALITY% -s %RESOLUTION% -f mpegts -c:v libx264 -c:a libmp3lame -maxrate %MAXBITRATE%k -preset superfast -threads 0 pipe:1"
encode_args_ogv = "-codec:v libtheora -qscale:v 7 -codec:a libvorbis -qscale:a 5 -f ogg pipe:1"
encode_get_image = "-ss %TIME% -f image2 -vframes 1 pipe:1"
encode_srt = "-vf \"subtitles='%SRTFILE%'\""
encode_ss_frame = "-ss %TIME%"
encode_ss_duration = "-t %DURATION%"
common_abbr = "divx,xvid,dvdrip,hdtv,lol,axxo,repack,xor,pdtv,real,vtv,caph,2hd,proper,fqm,uncut,topaz,tvt,notv,fpn,fov,orenji,0tv,omicron,dsr,ws,sys,crimson,wat,hiqt,internal,brrip,boheme,vost,vostfr,fastsub,addiction,x264,LOL,720p,1080p,YIFY,evolve,fihtv,first,bokutox,bluray,tvboom,info"

# Description

Attempt to access ampache subsonic API from client that uses the http GET pattern produces a 404, ie:
mysubsonicwebsite.com/rest/ping?u=user&t=d5b55a5c7191425fe06c3f1f30caa653&s=NVH41Q5EB&v=1.16.0&c=pSub&f=json

# Logs

No relevant logs from ampache

Apache log: 127.0.0.1 - - [02/Jun/2018:21:37:12 +0200] "GET /rest/ping?u=user&t=00dc368b8b9b60f914d0196a40a8295b&s=HJ9XNKfsA&v=1.16.0&c=pSub&f=json HTTP/1.0" 404 473 "-" "python-requests/2.18.4"

**Comment by wagnered** [comment]
@KimTheFirst ,

Ampache doesn't support the token/salt authentication: only username and plain text password.  The Subsonic API was a late add-on to Ampache and would have taken a major rework of the existing architecture. to include both hashed and non-hashed password.

Sorry about that...

**Comment by KimTheFirst** [comment]
I think the 404 is also because the default rewrite rules for both apache and nginx do not properly take into account the querystring; they are looking for a request ending in ".view" and if there is no such string in the request, they do not match and the server return a 404. Even if the client supports fallback, it never gets this far because the request never gets punted to rest/index.php.

Appending the following lines:
>RewriteRule ^index\.php$ -  [L]
>RewriteRule ^(.+)$ /rest/index.php?ssaction=$1 [L,QSA]

To the rest/.htaccess results in the request reaching index.php, and throwing the token authentication error.

At this point I created a SHA256 hash to use as a username and set the password using a lookup table. Then I got "Requested client version is not supported", so I deleted the version check from index.php.

Now my app can communicate with the server, but is still unable to download songs (It just says "failed to open file") but I think this is now an issue with the app or the API versions being mismatched.

In any case, I think the rewrite change may help some other clients connect to the server even if there are still major issues around auth and API versions.
